### PR TITLE
Use container parser of stanza for logs collection

### PR DESCRIPTION
**Pull request by ChrsMark**
Fixes https://github.com/open-telemetry/opentelemetry-helm-charts/issues/1188.

**_This is a draft until the `v0.102.0` version of Collector is released, which will include the new `max_log_size` setting of the parser._** 

### Testing notes

1. `helm install daemonset ./opentelemetry-collector --set presets.logsCollection.enabled=true --set image.repository="otel/opentelemetry-collector-k8s" --set command.name="otelcol-k8s" --set mode=daemonset`
2. `k get cm daemonset-opentelemetry-collector-agent -o jsonpath='{.data.relay}`:
```
exporters:
  debug: {}
extensions:
  health_check:
    endpoint: ${env:MY_POD_IP}:13133
processors:
  batch: {}
  memory_limiter:
    check_interval: 5s
    limit_percentage: 80
    spike_limit_percentage: 25
receivers:
  filelog:
    exclude:
    - /var/log/pods/default_daemonset-opentelemetry-collector*_*/opentelemetry-collector/*.log
    include:
    - /var/log/pods/*/*/*.log
    include_file_name: false
    include_file_path: true
    operators:
    - id: container-parser
      max_log_size: 102400
      type: container
    retry_on_failure:
      enabled: true
    start_at: end
  jaeger:
    protocols:
      grpc:
        endpoint: ${env:MY_POD_IP}:14250
      thrift_compact:
        endpoint: ${env:MY_POD_IP}:6831
      thrift_http:
        endpoint: ${env:MY_POD_IP}:14268
  otlp:
    protocols:
      grpc:
        endpoint: ${env:MY_POD_IP}:4317
      http:
        endpoint: ${env:MY_POD_IP}:4318
  prometheus:
    config:
      scrape_configs:
      - job_name: opentelemetry-collector
        scrape_interval: 10s
        static_configs:
        - targets:
          - ${env:MY_POD_IP}:8888
  zipkin:
    endpoint: ${env:MY_POD_IP}:9411
service:
  extensions:
  - health_check
  pipelines:
    logs:
      exporters:
      - debug
      processors:
      - memory_limiter
      - batch
      receivers:
      - otlp
      - filelog
    metrics:
      exporters:
      - debug
      processors:
      - memory_limiter
      - batch
      receivers:
      - otlp
      - prometheus
    traces:
      exporters:
      - debug
      processors:
      - memory_limiter
      - batch
      receivers:
      - otlp
      - jaeger
      - zipkin
  telemetry:
    metrics:
      address: ${env:MY_POD_IP}:8888
```

### CODE REVIEW

**Comment by ChrsMark** [comment]
Tested e2e as well on `kind v0.20.0 go1.20.4 linux/amd64`, k8s `v1.27.3`, runtime `containerd`. 
Sample records:

```console
2024-05-22T09:35:07.700Z	info	LogsExporter	{"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 2}
2024-05-22T09:35:07.700Z	info	ResourceLog #0
Resource SchemaURL: 
ScopeLogs #0
ScopeLogs SchemaURL: 
InstrumentationScope  
LogRecord #0
ObservedTimestamp: 2024-05-22 09:35:07.572874147 +0000 UTC
Timestamp: 2024-05-22 09:35:07.383200464 +0000 UTC
SeverityText: 
SeverityNumber: Unspecified(0)
Body: Str(otel logs at 09:35:07)
Attributes:
     -> log.file.path: Str(/var/log/pods/default_daemonset-logs-kk9m8_d5f6af9d-6ff6-4ba4-a8bb-dab9fea6838c/busybox/0.log)
     -> k8s.container.name: Str(busybox)
     -> k8s.namespace.name: Str(default)
     -> k8s.pod.name: Str(daemonset-logs-kk9m8)
     -> k8s.container.restart_count: Str(0)
     -> log.iostream: Str(stdout)
     -> logtag: Str(F)
     -> k8s.pod.uid: Str(d5f6af9d-6ff6-4ba4-a8bb-dab9fea6838c)
     -> time: Str(2024-05-22T09:35:07.383200464Z)
Trace ID: 
Span ID: 
Flags: 0
LogRecord #1
ObservedTimestamp: 2024-05-22 09:35:07.572974215 +0000 UTC
Timestamp: 2024-05-22 09:35:07.485134813 +0000 UTC
SeverityText: 
SeverityNumber: Unspecified(0)
Body: Str(otel logs at 09:35:07)
Attributes:
     -> k8s.pod.name: Str(daemonset-logs-kk9m8)
     -> k8s.pod.uid: Str(d5f6af9d-6ff6-4ba4-a8bb-dab9fea6838c)
     -> log.file.path: Str(/var/log/pods/default_daemonset-logs-kk9m8_d5f6af9d-6ff6-4ba4-a8bb-dab9fea6838c/busybox/0.log)
     -> time: Str(2024-05-22T09:35:07.485134813Z)
     -> log.iostream: Str(stdout)
     -> logtag: Str(F)
     -> k8s.namespace.name: Str(default)
     -> k8s.container.restart_count: Str(0)
     -> k8s.container.name: Str(busybox)
Trace ID: 
Span ID: 
Flags: 0
	{"kind": "exporter", "data_type": "logs", "name": "debug"}
2024-05-22T09:35:07.901Z	info	ResourceLog #0
Resource SchemaURL: 
ScopeLogs #0
ScopeLogs SchemaURL: 
InstrumentationScope  
LogRecord #0
ObservedTimestamp: 2024-05-22 09:35:07.772023047 +0000 UTC
Timestamp: 2024-05-22 09:35:07.587542234 +0000 UTC
SeverityText: 
SeverityNumber: Unspecified(0)
Body: Str(otel logs at 09:35:07)
Attributes:
     -> k8s.pod.name: Str(daemonset-logs-kk9m8)
     -> k8s.container.restart_count: Str(0)
     -> k8s.pod.uid: Str(d5f6af9d-6ff6-4ba4-a8bb-dab9fea6838c)
     -> k8s.container.name: Str(busybox)
     -> log.file.path: Str(/var/log/pods/default_daemonset-logs-kk9m8_d5f6af9d-6ff6-4ba4-a8bb-dab9fea6838c/busybox/0.log)
     -> log.iostream: Str(stdout)
     -> time: Str(2024-05-22T09:35:07.587542234Z)
     -> k8s.namespace.name: Str(default)
     -> logtag: Str(F)
Trace ID: 
Span ID: 
Flags: 0
LogRecord #1
ObservedTimestamp: 2024-05-22 09:35:07.772149911 +0000 UTC
Timestamp: 2024-05-22 09:35:07.689592419 +0000 UTC
SeverityText: 
SeverityNumber: Unspecified(0)
Body: Str(otel logs at 09:35:07)
Attributes:
     -> k8s.pod.uid: Str(d5f6af9d-6ff6-4ba4-a8bb-dab9fea6838c)
     -> k8s.container.name: Str(busybox)
     -> k8s.namespace.name: Str(default)
     -> k8s.pod.name: Str(daemonset-logs-kk9m8)
     -> time: Str(2024-05-22T09:35:07.689592419Z)
     -> log.iostream: Str(stdout)
     -> logtag: Str(F)
     -> log.file.path: Str(/var/log/pods/default_daemonset-logs-kk9m8_d5f6af9d-6ff6-4ba4-a8bb-dab9fea6838c/busybox/0.log)
     -> k8s.container.restart_count: Str(0)
Trace ID: 
Span ID: 
Flags: 0
```

File: charts/opentelemetry-collector/templates/_config.tpl [hunk 1]
```diff
@@ -204,82 +204,9 @@ receivers:
...
-        combine_with: ""
-        max_log_size: {{ $.Values.presets.logsCollection.maxRecombineLogSize }}
-      # Parse CRI-Containerd format
-      - type: regex_parser
-        id: parser-containerd
-        regex: '^(?P<time>[^ ^Z]+Z) (?P<stream>stdout|stderr) (?P<logtag>[^ ]*) ?(?P<log>.*)$'
-        timestamp:
-          parse_from: attributes.time
-          layout: '%Y-%m-%dT%H:%M:%S.%LZ'
-      - type: recombine
```
[the remark this answers is not in the document]
Comment (ChrsMark) [reply charts/opentelemetry-collector/templates/_config.tpl:240]: Yeap it [supports](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/docs/operators/container.md#parse-the-multiline-as-containerd-container-log-and-recombine-into-a-single-one) that already.

File: charts/opentelemetry-collector/templates/_config.tpl [hunk 2]
```diff
@@ -204,82 +204,9 @@ receivers:
...
-        id: containerd-recombine
-        output: extract_metadata_from_filepath
-        combine_field: attributes.log
-        source_identifier: attributes["log.file.path"]
-        is_last_entry: "attributes.logtag == 'F'"
-        combine_with: ""
-        max_log_size: {{ $.Values.presets.logsCollection.maxRecombineLogSize }}
```
[the remark this answers is not in the document]
Comment (ChrsMark) [reply charts/opentelemetry-collector/templates/_config.tpl:247]: Good point, `max_log_size` is set internally to `102400`. We will need to make this configurable in the parser.

Let's stall this PR then, until we provide the support for this.

**Comment by ChrsMark** [comment]
This should not get merged until the `v0.102.0` version of Collector is released, which will include the new max_log_size setting of the parser.

**Comment by TylerHelmuth** [comment]
@ChrsMark `v0.102.1` is out, this is ready to be picked up again

**Comment by ChrsMark** [comment]
There was an additional fix for the parser that did not make it to `v0.102.0`: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33353

I would suggest we wait for it, before doing the switch.

**Comment by ChrsMark** [comment]
Not stale.

**Comment by ChrsMark** [comment]
@TylerHelmuth finally the fixes are in. This one should be ready for a review now.

### Test the produced configmap

1. Run `helm install daemonset ./charts/opentelemetry-collector --set presets.logsCollection.enabled=true --set image.repository="otel/opentelemetry-collector-k8s" --set command.name="otelcol-k8s" --set mode=daemonset --dry-run`

2. Verify the produced config:
```yaml
    receivers:
      filelog:
        exclude:
        - /var/log/pods/default_daemonset-opentelemetry-collector*_*/opentelemetry-collector/*.log
        include:
        - /var/log/pods/*/*/*.log
        include_file_name: false
        include_file_path: true
        operators:
        - id: container-parser
          max_log_size: 102400
          type: container
        retry_on_failure:
          enabled: true
        start_at: end
```

### Verify the produced log records

1. `helm install daemonset ./charts/opentelemetry-collector --set presets.logsCollection.enabled=true --set image.repository="otel/opentelemetry-collector-k8s" --set command.name="otelcol-k8s" --set mode=daemonset --set config.exporters.debug.verbosity="detailed"`

2. Verify that Resource attribute and log attributes are populated properly:
```console
2024-06-25T08:15:12.981Z	info	ResourceLog #0
Resource SchemaURL: 
Resource attributes:
     -> k8s.namespace.name: Str(default)
     -> k8s.pod.name: Str(daemonset-logs-8w7rm)
     -> k8s.container.restart_count: Str(0)
     -> k8s.pod.uid: Str(539b21bf-0b23-4929-ad61-a5c76820b309)
     -> k8s.container.name: Str(busybox)
ScopeLogs #0
ScopeLogs SchemaURL: 
InstrumentationScope  
LogRecord #0
ObservedTimestamp: 2024-06-25 08:15:12.739110323 +0000 UTC
Timestamp: 2024-06-25 08:15:12.606659689 +0000 UTC
SeverityText: 
SeverityNumber: Unspecified(0)
Body: Str(otel logs at 08:15:12)
Attributes:
     -> logtag: Str(F)
     -> log.file.path: Str(/var/log/pods/default_daemonset-logs-8w7rm_539b21bf-0b23-4929-ad61-a5c76820b309/busybox/0.log)
     -> time: Str(2024-06-25T08:15:12.606659689Z)
     -> log.iostream: Str(stdout)
Trace ID: 
Span ID: 
Flags: 0
LogRecord #1
ObservedTimestamp: 2024-06-25 08:15:12.739202433 +0000 UTC
Timestamp: 2024-06-25 08:15:12.707975372 +0000 UTC
SeverityText: 
SeverityNumber: Unspecified(0)
Body: Str(otel logs at 08:15:12)
Attributes:
     -> log.file.path: Str(/var/log/pods/default_daemonset-logs-8w7rm_539b21bf-0b23-4929-ad61-a5c76820b309/busybox/0.log)
     -> time: Str(2024-06-25T08:15:12.707975372Z)
     -> log.iostream: Str(stdout)
     -> logtag: Str(F)
Trace ID: 
Span ID: 
Flags: 0
	{"kind": "exporter", "data_type": "logs", "name": "debug"}
```

File: charts/opentelemetry-collector/templates/_config.tpl [hunk 3]
```diff
@@ -204,82 +204,10 @@ receivers:
...
-      - type: recombine
-        id: crio-recombine
-        output: extract_metadata_from_filepath
-        combine_field: attributes.log
-        source_identifier: attributes["log.file.path"]
-        is_last_entry: "attributes.logtag == 'F'"
-        combine_with: ""
+      # parse container logs
+      - type: container
+        id: container-parser
```
Comment (TylerHelmuth) [inline charts/opentelemetry-collector/templates/_config.tpl:209]: @ChrsMark @djaglowski thanks for making this parser, its amazing.

**Review by TylerHelmuth — APPROVED** [review]

**Pull request merged**